lib/still/data.ex

```
@@ -0,0 +1,162 @@
+  Still will consider the pages inside as web pages or assets to build.
+
+  It supports JSON (.json), YAML (.yml) and Elixir (.exs, .ex).
+  """
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
+  end
+
+  def load do
+    GenServer.call(__MODULE__, :load, :infinity)
+  end
+
+  def global do
+    GenServer.call(__MODULE__, :global, :infinity)
+  end
+
+  def reset do
+    GenServer.call(__MODULE__, :reset, :infinity)
+  end
+
+  def member?(input_file) do
+    String.starts_with?(input_file, data_folder())
+  end
+
+  @impl true
+  def init(_) do
+    {:ok, @initial_state}
+  end
+
+  @impl true
+  def handle_call(:reset, _, _state) do
+    {:reply, :ok, @initial_state}
+  end
+
+  @impl true
+  def handle_call(:global, _, state) do
+    {:reply, state.global, state}
+  end
+
+  @impl true
+  def handle_call(:load, _, state) do
+    data =
+      data_files()
+      |> sort_files()
+      |> Enum.reduce(%{}, &load_file/2)
+
+    {:reply, :ok, %{state | global: data}}
+  end
+
+  defp data_files(rel_path \\ "") do
+    path =
+      data_folder()
+      |> get_input_path()
+      |> Path.join(rel_path)
+
+    cond do
+      File.regular?(path) && Enum.member?(@supported_extensions, Path.extname(path)) ->
+        [rel_path]
+
+      File.dir?(path) ->
+        File.ls!(path)
+        |> Enum.map(&data_files(Path.join(rel_path, &1)))
+
+      true ->
+        []
+    end
+    |> List.flatten()
+  end
+
+  defp sort_files(files) do
+    files
+    |> Enum.sort_by(fn file ->
+      file
+      |> path_components()
+      |> length()
+    end)
+  end
+
+  defp load_file(input_file, data) do
+    data_folder()
+    |> Path.join(input_file)
+    |> get_input_path()
+    |> read_file()
+    |> case do
+      {:ok, result} ->
+        put_in(
+          data,
+          Enum.map(path_components(input_file), &Access.key(&1, %{})),
+          result
+        )
+
+      {:error, err} ->
+        raise err
```

Where's this being caught? Should we raise a custom error instead?

---

This was for Jason.decode. I changed it to Jason.decode! because we really want to see that error if something's wrong.